crates/ruff_db/src/diagnostic.rs

```diff
@@ -152,7 +152,7 @@ pub trait Diagnostic: Send + Sync + std::fmt::Debug {
 
     fn message(&self) -> Cow<str>;
 
-    fn file(&self) -> File;
+    fn file(&self) -> Option<File>;
```

Comment:
is it worth adding a doc-comment explaining why `file` might sometimes be `None`? Your explanation in the PR summary is very good!